lib/pengine/rules.c

```diff
@@ -8,15 +8,17 @@
  */
 
 #include <crm_internal.h>
-#include <crm/crm.h>
-#include <crm/common/xml.h>
-#include <crm/common/xml_internal.h>
 
 #include <glib.h>
 
+#include <crm/crm.h>
+#include <crm/common/xml.h>
 #include <crm/pengine/rules.h>
-#include <crm/pengine/rules_internal.h>
+
+#include <crm/common/iso8601_internal.h>
```

Comment:
The only technical requirement is that crm_internal.h must be first, because it modifies the behavior of some system headers (defining allowed API versions for glib, and defining `_GNU_SOURCE` via portability.h to change glibc behavior).

I lean to all non-Pacemaker headers after that, then all Pacemaker headers, but I don't have a preference for the order within each of those groups. It *shouldn't* matter for Pacemaker headers, since each should include everything it needs (make check ensures that for public headers -- we should probably extend that test to internal headers). Most system headers do too, but some don't and require you to include something else first (often sys/types.h or something like that), so we can't establish a firm pattern for those.

Reply:
My eventual goal is to ensure that we only need to include one public header and one internal header per Pacemaker library, to make the question mostly moot. It would be `crm/crm.h` and `crm/crm_internal.h` for libcrmcommon, to minimize the impact on external code.

Lately I've been leaning to this idea:
* Every source file in a library would have corresponding headers for the APIs defined in it, as `include/.../BASENAME{,_internal}.h` and `lib/.../BASENAME_private.h`.
* The top-level headers for each library would contain nothing but including all the other headers of the appropriate type.
* The top-level headers would define a constant before including the other headers, and undefine it after, and the subheaders would `#error` if that constant was not defined (to prevent the subheaders from being included by themselves and ensure that only the toplevel headers can be directly included). This could be done immediately for internal/private headers but public headers would need some leeway for existing external code. (Of course we would stop requiring subheaders to be includable alone at this point.)

Reply:
Created T777 for that